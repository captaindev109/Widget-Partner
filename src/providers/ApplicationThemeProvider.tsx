import { ThemeProvider, DefaultTheme } from 'styled-components';
import GlobalStyles from '@/styles/Global';
import React from 'react';
import { ChildrenReturnType } from '@/types/types';

interface ApplicationThemeProviderProps {
  children?: ChildrenReturnType;
  theme: DefaultTheme;
}

const ApplicationThemeProvider = ({ children, theme }: ApplicationThemeProviderProps) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};

export default ApplicationThemeProvider;
