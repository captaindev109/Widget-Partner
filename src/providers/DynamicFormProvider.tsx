import React from 'react';
import { ChildrenReturnType } from '@/types/types';
import { FormProvider, useForm } from 'react-hook-form';

interface DynamicFormProviderProps {
  children?: ChildrenReturnType;
}

const useDynamicRootForm = () => {
  const methods = useForm({ mode: 'onChange' });

  return { methods };
};

export const DynamicFormProvider = ({ children }: DynamicFormProviderProps) => {
  const { methods } = useDynamicRootForm();

  return <FormProvider {...methods}>{children}</FormProvider>;
};
