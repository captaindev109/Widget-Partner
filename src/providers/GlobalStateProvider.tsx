import { mainMachine } from '@/machines/courseMachine';
import headerMachine from '@/machines/headerMachine';
import footerMachine from '@/machines/footerMachine';
import React, { createContext } from 'react';
import { useInterpret } from '@xstate/react';
import { ActorRefFrom } from 'xstate';
import { storeMachine } from '@/machines/storeMachine';

interface GlobalStateContextProps {
  mainService: ActorRefFrom<typeof mainMachine>;
  headerService: ActorRefFrom<typeof headerMachine>;
  footerService: ActorRefFrom<typeof footerMachine>;
  storeService: ActorRefFrom<typeof storeMachine>;
}

export const GlobalStateContext = createContext({} as GlobalStateContextProps);

interface GlobalStateProviderProps {
  children?: React.ReactNode;
}

const useDevTool = process.env.NODE_ENV === 'development';

export const GlobalStateProvider = ({ children }: GlobalStateProviderProps) => {
  const services = {
    mainService: useInterpret(mainMachine, {
      devTools: useDevTool,
    }),
    headerService: useInterpret(headerMachine, {
      devTools: useDevTool,
    }),
    footerService: useInterpret(footerMachine, {
      devTools: useDevTool,
    }),
    storeService: useInterpret(storeMachine, {
      devTools: useDevTool,
    }),
  };

  return <GlobalStateContext.Provider value={services}>{children}</GlobalStateContext.Provider>;
};
