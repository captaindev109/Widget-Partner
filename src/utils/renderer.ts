import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
