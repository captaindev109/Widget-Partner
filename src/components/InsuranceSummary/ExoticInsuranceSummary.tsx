import useEditQuote from '@/hooks/useEditQuote';
import useQuoteStore from '@/hooks/useQuoteStore';
import useStore from '@/hooks/useStore';
import { parseData } from '@/utils/dataFormatters';
import { FC } from 'react';
import InsuranceSummary from '.';
import useDynamicForm from '../DynamicForm/useDynamicForm';

const ExoticInsuranceSummary: FC = () => {
  const { userSelector, formDataSelector, updateStoreUser } = useStore();
  const { handleEditQuotes } = useEditQuote();
  const { rehydrateQuoteCart } = useQuoteStore();
  const { submitForm, resetForm } = useDynamicForm();

  const handleSave = async () => {
    const formData = await submitForm();
    if (formData) {
      await handleEditQuotes(formData);
      await rehydrateQuoteCart();

      // Update user with the form data
      const userData = parseData({ data_to_parse: formData, type_of_parsing: 'form_data_to_user' });
      updateStoreUser(userData);
    }
  };

  const handleClose = async () => {
    await resetForm();
  };

  return <InsuranceSummary user={userSelector} data={formDataSelector} onSave={handleSave} onClose={handleClose} />;
};

export default ExoticInsuranceSummary;
