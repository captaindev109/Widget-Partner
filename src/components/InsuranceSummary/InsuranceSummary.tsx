import { FC, useRef, useState } from 'react';
import styled from 'styled-components';
import { InsuranceSummaryProps, EachComponentDataSchema } from '@/types/types';
import { List } from '../List';
import Card from '@/generics/Card';
import Text from '@/generics/Text';
import { ModalRef } from '../Modal/Modal';
import Modal from '../Modal';
import Button from '../Button';
import View from '@/generics/View';
import viewRenderer from '@/utils/renderer';
import { device, HOME_INSURANCE_TYPES } from '@/utils/constants';

const ModalContentWrapper = styled(Card)`
  display: flex;
  flex-direction: column;
  padding: 0px 16px 16px;

  width: 90vw;
  height: 90vh;

  @media ${device.tablet} {
    padding: 0px 32px 32px;
    width: 70vw;
    height: 70vh;
  }
`;

const Divider = styled(Card)`
  border-bottom: 1px solid #eaeaea;
`;

//Information Modal Json Data
const ComponentsDataSchema = [
  {
    id: 'information_modal_checkout',
    type: 'display',
    component: 'Card',
    children: [
      {
        id: 'title_card_0001',
        component: 'Card',
        marginBottom: '24px',
        children: [
          {
            id: 'card_0001',
            component: 'Text',
            variant: 'HeadlineS',
            children: 'Ändra uppgifter',
          },
        ],
      },
      {
        id: 'about_you_form',
        component: 'DynamicForm',
        children: [
          {
            id: 'birthDate',
            component: 'Input',
            validationRules: {
              required: {
                value: true,
                message: 'Du måste ange födelsedatum',
              },
              pattern: {
                value: /^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01])$/,
                message: 'Ange ett giltigt födelsedatum (YYYY-MM-DD)',
              },
            },
            label: 'Födelsedatum',
            name: 'birthDate',
            type: 'text',
            placeholder: '1994-11-15',
          },
          {
            id: 'numberOfPeople',
            component: 'Dropdown',
            label: 'Välj antal personer att försäkra',
            name: 'householdSize',
            placeholder: '1',
            validationRules: {
              required: {
                value: true,
                message: 'Du måste ange storlek',
              },
            },
            tooltip: 'Välj antal personer att försäkra, om du vill försäkra fler än en person, välj fler personer.',
            options: [
              { label: '1', value: 1 },
              { label: '2', value: 2 },
              { label: '3', value: 3 },
              { label: '4', value: 4 },
              { label: '5', value: 5 },
              { label: '6', value: 6 },
            ],
          },
          {
            id: 'student',
            component: 'Dropdown',
            label: 'Är du student?',
            name: 'student',
            placeholder: 'No',
            validationRules: {
              required: {
                value: true,
                message: 'Du måste ange om du är student',
              },
            },
            options: [
              { label: 'No', value: 'no' },
              { label: 'Yes', value: 'yes' },
            ],
          },
          {
            id: 'address',
            component: 'Input',
            validationRules: {
              required: {
                value: true,
                message: 'Du måste ange gatuadress',
              },
              pattern: {
                value: '^(.+),([^,]+),([^,]+)$',
                message: 'Fel format (Gata, Postnummer, Stad)',
              },
            },
            label: 'Gatuadress',
            name: 'address',
            type: 'text',
            placeholder: 'Vagnvägen 4, 12345 Västerås',
          },
          {
            id: 'size',
            component: 'Input',
            validationRules: {
              required: {
                value: true,
                message: 'Du måste ange storlek',
              },
            },
            label: 'Boarea',
            name: 'livingSpace',
            type: 'number',
            placeholder: '34 m2',
          },
          {
            id: 'subType',
            component: 'Dropdown',
            label: 'Typ av bostad',
            name: 'subType',
            placeholder: 'RENT',
            validationRules: {
              required: {
                value: true,
                message: 'Du måste ange typ av bostad',
              },
            },
            options: [
              { label: 'Hyresrätt', value: 'RENT' },
              { label: 'Bostadsrätt', value: 'BRF' },
            ],
          },
        ],
      },
    ],
  },
];

//The main component with dynamic rendering and state
const InsuranceSummary: FC<InsuranceSummaryProps> = ({ onSave, onClose, user, data }) => {
  const modalRef = useRef<ModalRef>(null);
  const [error, setError] = useState<boolean>(false);
  const [loading, setLoading] = useState<boolean>(false);
  const { firstName, lastName, birthDate, email, phoneNumber } = user;

  const handleOpenModal = () => {
    modalRef.current?.toggle?.();
  };

  const handleSave = async () => {
    try {
      setError(false);
      setLoading(true);
      await onSave?.();
      modalRef.current?.toggle?.();
    } catch (err) {
      setError(true);
    } finally {
      setLoading(false);
    }
  };

  const handleClose = async () => {
    await onClose?.();
  };

  return (
    <Card shadow={1} borderRadius="8px" padding="16px 18px" background="white">
      <List data-testid="InsuranceSummary" display="flex" direction="column" spaceBetween="16px">
        <Card display="flex" justifyContent="space-between">
          <Text variant="Body3" color="gray900">
            Försäkringstagare
          </Text>
          <Text variant="Body3" color="gray700">
            {firstName} {lastName}
          </Text>
        </Card>
        <Card display="flex" justifyContent="space-between">
          <Text variant="Body3" color="gray900">
            Antal försäkrade
          </Text>
          <Text variant="Body3" color="gray700">
            {data.householdSize > 1 ? ` Du + ${data.householdSize - 1} personer` : '1 person'}
          </Text>
        </Card>
        <Card display="flex" justifyContent="space-between">
          <Text variant="Body3" color="gray900">
            Födelsedag
          </Text>
          <Text variant="Body3" color="gray700">
            {birthDate}
          </Text>
        </Card>
        <Divider />
        <Card display="flex" justifyContent="space-between">
          <Text variant="Body3" color="gray900">
            Gatuadress
          </Text>
          <Text variant="Body3" color="gray700">
            {data.address}
          </Text>
        </Card>
        {/* <Card display="flex" justifyContent="space-between">
          <Text variant="Body3" color="gray900">
            Postkod
          </Text>
          <Text variant="Body3" color="gray700">
            {data.zipCode}
          </Text>
        </Card> */}
        <Card display="flex" justifyContent="space-between">
          <Text variant="Body3" color="gray900">
            Typ av bostad
          </Text>
          <Text variant="Body3" color="gray700">
            {data?.subType && HOME_INSURANCE_TYPES[data?.subType]}
          </Text>
        </Card>
        <Divider />
        <Card display="flex" justifyContent="space-between">
          <Text variant="Body3" color="gray900">
            Telefon
          </Text>
          <Text variant="Body3" color="gray700">
            {phoneNumber}
          </Text>
        </Card>
        <Card display="flex" justifyContent="space-between">
          <Text variant="Body3" color="gray900">
            Epost
          </Text>
          <Text variant="Body3" color="gray700">
            {email}
          </Text>
        </Card>
        <Divider />
      </List>
      <Button variant="transparent" onClick={handleOpenModal}>
        <Text variant="Body3" color="purple900">
          Ändra uppgifter
        </Text>
      </Button>
      <Modal ref={modalRef} customCloseMethod={handleClose}>
        <ModalContentWrapper>
          <View components={ComponentsDataSchema}>
            {(component: EachComponentDataSchema) => viewRenderer(component, null)}
          </View>
          <Card width="107px" margin="auto" padding="32px 0px 16px 0px">
            <Button disabled={loading} variant="primary" onClick={handleSave}>
              Spara
            </Button>
          </Card>
          <Card textAlign="center" paddingBottom="32px">
            {error ? (
              <Text color="red500">Det gick tyvärr inte att spara, se till att all information är korrekt.</Text>
            ) : (
              <Text>Ditt pris kan komma att ändras</Text>
            )}
          </Card>
        </ModalContentWrapper>
      </Modal>
    </Card>
  );
};

export default InsuranceSummary;
