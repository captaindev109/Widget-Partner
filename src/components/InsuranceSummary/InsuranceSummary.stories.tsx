import React from 'react';
import { Meta, Story } from '@storybook/react';
import styled from 'styled-components';
import { decorators } from '@/../.storybook/preview';
import InsuranceSummary from '.';
import { InsuranceSummaryProps } from '@/types/types';
import Wrapper from '@/generics/Wrapper';

// A wrapper to position the component
const StoryWrapper = styled(Wrapper)``;

// Name the folder to either generics or components
const meta: Meta = {
  title: 'Components/InsuranceSummary',
  component: InsuranceSummary,
  decorators,
};

export default meta;

// Showcasing the component
const Template: Story<InsuranceSummaryProps> = (args) => (
  <StoryWrapper>
    <InsuranceSummary {...args} />
  </StoryWrapper>
);

export const DefaultInsuranceSummary = Template.bind({});

DefaultInsuranceSummary.args = {
  user: {
    firstName: 'Magnus',
    lastName: 'Berg',
    birthDate: '1996-05-17',
    email: 'magnus.berg.66@gmail.com',
  },
  data: {
    street: 'Vagnvagen 4',
    numberOfPeople: 4,
    phoneNumber: '+46705309084',
    typeOfResidence: 'Rental',
    zipCode: '175 67',
  },
};
