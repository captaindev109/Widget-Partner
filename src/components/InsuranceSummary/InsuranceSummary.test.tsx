/**
 *
 *  ********** Unit & Snapshot Testing *********
 *
 * Unit test is specifiic for each functionliaty of the component
 * Snapshot test is to track Over All changes in the code since the first snap
 **/

import { render, RenderResult } from '@testing-library/react';
import queries from '@testing-library/dom/types/queries';
import renderer from 'react-test-renderer';
import InsuranceSummary from './InsuranceSummary';
import { GlobalStateProvider } from '@/providers/GlobalStateProvider';
import ApplicationThemeProvider from '@/providers/ApplicationThemeProvider';
import { DynamicFormProvider } from '@/providers/DynamicFormProvider';
import defaultTheme from '@/theme/theme';

const data = {
  user: {
    firstName: 'Magnus',
    lastName: 'Berg',
    birthDate: '12/24/1991',
    email: 'magnus.berg.66@gmail.com',
  },
  data: {
    street: 'Vagnvagen 4',
    numberOfPeople: 4,
    phoneNumber: '+46705309084',
    typeOfResidence: 'Rental',
    zipCode: '175 67',
  },
};

const MockComponent: React.FC = () => {
  return (
    <GlobalStateProvider>
      <ApplicationThemeProvider theme={defaultTheme}>
        <DynamicFormProvider>
          <InsuranceSummary {...data} />
        </DynamicFormProvider>
      </ApplicationThemeProvider>
    </GlobalStateProvider>
  );
};

// describe function is a better way to organize our test cases and describe which component we are testing
describe('<InsuranceSummary /> Test Cases', () => {
  let renderedComponent: RenderResult<typeof queries, HTMLElement>;
  beforeEach(() => {
    renderedComponent = render(<MockComponent />);
  });

  // this test is a must for each component, any rendering issues with the component can be found
  test('it renders without crashing, all is well', () => {
    // render basically prints and exposes the component in nodejs environemt, it returns an object...
    render(<MockComponent />);
  });

  test('check to see any specific string, element in dom, for example', () => {
    const { container } = renderedComponent;

    //using selectors
    expect(container.querySelector('ul')).toBeInTheDocument();
  });

  it('Snapshot test should pass', () => {
    const snap = renderer.create(<MockComponent />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
