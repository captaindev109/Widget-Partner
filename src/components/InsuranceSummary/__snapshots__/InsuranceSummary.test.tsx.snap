// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InsuranceSummary /> Test Cases Snapshot test should pass 1`] = `
<div
  className="Card-sc-fb64gk-0 CWoXT"
>
  <ul
    className="List__StyledList-sc-1b1or9q-0 lfyYCb"
    direction="column"
    display="flex"
  >
    <div
      className="Card-sc-fb64gk-0 gZFGrF"
      display="flex"
    >
      <p
        className="Text__StyledText-sc-1jfsuh1-0 kuEDKQ"
        color="gray900"
      >
        Försäkringstagare
      </p>
      <p
        className="Text__StyledText-sc-1jfsuh1-0 lmAVHO"
        color="gray700"
      >
        Magnus
         
        Berg
      </p>
    </div>
    <div
      className="Card-sc-fb64gk-0 gZFGrF"
      display="flex"
    >
      <p
        className="Text__StyledText-sc-1jfsuh1-0 kuEDKQ"
        color="gray900"
      >
        Antal försäkrade
      </p>
      <p
        className="Text__StyledText-sc-1jfsuh1-0 lmAVHO"
        color="gray700"
      >
        1 person
      </p>
    </div>
    <div
      className="Card-sc-fb64gk-0 gZFGrF"
      display="flex"
    >
      <p
        className="Text__StyledText-sc-1jfsuh1-0 kuEDKQ"
        color="gray900"
      >
        Födelsedag
      </p>
      <p
        className="Text__StyledText-sc-1jfsuh1-0 lmAVHO"
        color="gray700"
      >
        12/24/1991
      </p>
    </div>
    <div
      className="Card-sc-fb64gk-0 InsuranceSummary__Divider-sc-18vjd4x-1 leCJCh hwYbks"
    />
    <div
      className="Card-sc-fb64gk-0 gZFGrF"
      display="flex"
    >
      <p
        className="Text__StyledText-sc-1jfsuh1-0 kuEDKQ"
        color="gray900"
      >
        Gatuadress
      </p>
      <p
        className="Text__StyledText-sc-1jfsuh1-0 lmAVHO"
        color="gray700"
      />
    </div>
    <div
      className="Card-sc-fb64gk-0 gZFGrF"
      display="flex"
    >
      <p
        className="Text__StyledText-sc-1jfsuh1-0 kuEDKQ"
        color="gray900"
      >
        Typ av bostad
      </p>
      <p
        className="Text__StyledText-sc-1jfsuh1-0 lmAVHO"
        color="gray700"
      />
    </div>
    <div
      className="Card-sc-fb64gk-0 InsuranceSummary__Divider-sc-18vjd4x-1 leCJCh hwYbks"
    />
    <div
      className="Card-sc-fb64gk-0 gZFGrF"
      display="flex"
    >
      <p
        className="Text__StyledText-sc-1jfsuh1-0 kuEDKQ"
        color="gray900"
      >
        Telefon
      </p>
      <p
        className="Text__StyledText-sc-1jfsuh1-0 lmAVHO"
        color="gray700"
      />
    </div>
    <div
      className="Card-sc-fb64gk-0 gZFGrF"
      display="flex"
    >
      <p
        className="Text__StyledText-sc-1jfsuh1-0 kuEDKQ"
        color="gray900"
      >
        Epost
      </p>
      <p
        className="Text__StyledText-sc-1jfsuh1-0 lmAVHO"
        color="gray700"
      >
        magnus.berg.66@gmail.com
      </p>
    </div>
    <div
      className="Card-sc-fb64gk-0 InsuranceSummary__Divider-sc-18vjd4x-1 leCJCh hwYbks"
    />
  </ul>
  <button
    className="Button__StyledButton-sc-1drbzk-0 kPuKUj"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <div
      className="Card-sc-fb64gk-0 Button__ContentWrapper-sc-1drbzk-1 leCJCh jowrmu"
    >
      <p
        className="Text__StyledText-sc-1jfsuh1-0 fxAzWg"
        color="purple900"
      >
        Ändra uppgifter
      </p>
    </div>
  </button>
</div>
`;
